src/Storages/ObjectStorage/DataLakes/IcebergMetadata.cpp

```diff
@@ -240,91 +235,203 @@ DataTypePtr getFieldType(const Poco::JSON::Object::Ptr & field, const String & t
+{
+    bool allowed_type_conversion = (old_type == new_type);
+    allowed_type_conversion |= (old_type == "int") && (new_type == "long");
+    allowed_type_conversion |= (old_type == "float") && (new_type == "double");
+    if (old_type.starts_with("decimal(") && old_type.ends_with(')') && new_type.starts_with("decimal(") && new_type.ends_with(")"))
     {
-        current_schema_id = metadata_object->getValue<int>("current-schema-id");
-        auto schemas = metadata_object->get("schemas").extract<Poco::JSON::Array::Ptr>();
-        if (schemas->size() == 0)
-            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Cannot parse Iceberg table schema: schemas field is empty");
+        auto [old_precision, old_scale] = parseDecimal(old_type);
+        auto [new_precision, new_scale] = parseDecimal(new_type);
+        allowed_type_conversion |= (old_precision <= new_precision) && (old_scale == new_scale);
+    }
+    return allowed_type_conversion;
+}
 
-        if (ignore_schema_evolution)
+std::shared_ptr<ActionsDAG>
+IcebergSchemaProcessor::getSchemaTransformationDag(const Poco::JSON::Object::Ptr & old_schema, const Poco::JSON::Object::Ptr & new_schema)
+{
+    std::map<size_t, std::pair<Poco::JSON::Object::Ptr, const ActionsDAG::Node *>> old_schema_entries;
+    auto old_schema_fields = old_schema->get("fields").extract<Poco::JSON::Array::Ptr>();
+    std::shared_ptr<ActionsDAG> dag = std::make_shared<ActionsDAG>();
+    auto & outputs = dag->getOutputs();
+    for (size_t i = 0; i != old_schema_fields->size(); ++i)
+    {
+        auto field = old_schema_fields->getObject(static_cast<UInt32>(i));
+        size_t id = field->getValue<size_t>("id");
+        auto name = field->getValue<String>("name");
+        bool required = field->getValue<bool>("required");
+        old_schema_entries[id] = {field, &dag->addInput(name, getFieldType(field, "type", required))};
+    }
+    auto new_schema_fields = new_schema->get("fields").extract<Poco::JSON::Array::Ptr>();
+    for (size_t i = 0; i != new_schema_fields->size(); ++i)
+    {
+        auto field = new_schema_fields->getObject(static_cast<UInt32>(i));
+        size_t id = field->getValue<size_t>("id");
+        auto name = field->getValue<String>("name");
+        bool required = field->getValue<bool>("required");
+        auto type = getFieldType(field, "type", required);
+        if (old_schema_entries.contains(id))
         {
-            /// If we ignore schema evolution, we will just use latest schema for all data files.
-            /// Find schema with 'schema-id' equal to 'current_schema_id'.
-            for (uint32_t i = 0; i != schemas->size(); ++i)
+            auto [old_json, old_node] = old_schema_entries.find(id)->second;
+            if (field->isObject("type"))
             {
-                auto current_schema = schemas->getObject(i);
-                if (current_schema->getValue<int>("schema-id") == current_schema_id)
+                if (!(*old_json == *field))
```

Comment:
Let's maybe implement operator `!=` with `!` + `==` and use it here to make it look better?